File: options/rawpack.py
import os
import shutil
from typing import Callable, List, Tuple


from fs.rawpack import (
    get_num_set_file_names,
    move_out_files_in_folder_in_cache_dir,
    unzip_file_to_cache_dir,
)
from fs.move import is_dir_having_file, move_elements_across_dir
from options.base import InputType


def unzip_numeric_to_bms_folder(
    pack_dir: str, cache_dir: str, root_dir: str, confirm: bool = False
):
    if not os.path.isdir(cache_dir):
        os.mkdir(cache_dir)
    if not os.path.isdir(root_dir):
        os.mkdir(root_dir)

    num_set_file_names: List[str] = get_num_set_file_names(pack_dir)

    if confirm:
        for file_name in num_set_file_names:
            print(f" --> {file_name}")
        selection = input("-> Confirm [y/N]:")
        if selection.lower().startswith("y"):
            return

    for file_name in num_set_file_names:
        file_path = os.path.join(pack_dir, file_name)
        id_str = file_name.split(" ")[0]

        # Prepare an empty cache dir
        cache_dir_path = os.path.join(cache_dir, id_str)

        if os.path.isdir(cache_dir_path) and is_dir_having_file(cache_dir_path):
            shutil.rmtree(cache_dir_path)

        if not os.path.isdir(cache_dir_path):
            os.mkdir(cache_dir_path)

        # Unpack & Copy
        unzip_file_to_cache_dir(file_path, cache_dir_path)

        # Move files in dir
        move_result = move_out_files_in_folder_in_cache_dir(cache_dir_path)
        if not move_result:
            continue

        # Find Existing Target dir
        target_dir_path = None
        for dir_name in os.listdir(root_dir):
            dir_path = os.path.join(root_dir, dir_name)
            if not os.path.isdir(dir_path):
                continue
            if not (
                dir_name.startswith(id_str)
                and (
                    len(dir_name) == len(id_str)
                    or dir_name[len(id_str) :].startswith(".")
                )
            ):
                continue
            target_dir_path = dir_path

        # Create New Target dir
        if target_dir_path is None:
            target_dir_path = os.path.join(root_dir, id_str)

        if not os.path.isdir(target_dir_path):
            os.mkdir(target_dir_path)

        # Move cache to bms dir
        print(f" > Moving files in {cache_dir_path} to {target_dir_path}")
        move_elements_across_dir(cache_dir_path, target_dir_path)
        try:
            os.rmdir(cache_dir_path)
        except FileNotFoundError:
            pass

        # Move File to Another dir
        print(f" > Finish dealing with file: {file_name}")
        used_pack_dir = os.path.join(pack_dir, "BOFTTPacks")
        if not os.path.isdir(used_pack_dir):
            os.mkdir(used_pack_dir)
        shutil.move(file_path, os.path.join(used_pack_dir, file_name))


def unzip_with_name_to_bms_folder(
    pack_dir: str, cache_dir: str, root_dir: str, confirm: bool = False
):
    if not os.path.isdir(cache_dir):
        os.mkdir(cache_dir)
    if not os.path.isdir(root_dir):
        os.mkdir(root_dir)

    num_set_file_names: List[str] = [
        file_name
        for file_name in os.listdir(pack_dir)
        if os.path.isfile(os.path.join(pack_dir, file_name))
        and (
            file_name.endswith(".zip")
            or file_name.endswith(".7z")
            or file_name.endswith(".rar")
        )
    ]

    if confirm:
        for file_name in num_set_file_names:
            print(f" --> {file_name}")
        selection = input("-> Confirm [y/N]:")
        if selection.lower().startswith("y"):
            return

    for file_name in num_set_file_names:
        file_path = os.path.join(pack_dir, file_name)
        file_name_without_ext = file_name[: -len(file_name.rsplit(".", 1)[-1]) - 1]
        while len(file_name_without_ext) > 0 and file_name_without_ext[-1] == ".":
            file_name_without_ext = file_name_without_ext[:-1]

        # Prepare an empty cache dir
        cache_dir_path = os.path.join(cache_dir, file_name_without_ext)

        if os.path.isdir(cache_dir_path) and is_dir_having_file(cache_dir_path):
            shutil.rmtree(cache_dir_path)

        if not os.path.isdir(cache_dir_path):
            os.mkdir(cache_dir_path)

        # Unpack & Copy
        unzip_file_to_cache_dir(file_path, cache_dir_path)

        # Move files in dir
        move_result = move_out_files_in_folder_in_cache_dir(cache_dir_path)
        if not move_result:
            continue

        target_dir_path = os.path.join(root_dir, file_name_without_ext)

        # Create New Target dir
        if not os.path.isdir(target_dir_path):
            os.mkdir(target_dir_path)

        # Move cache to bms dir
        print(f" > Moving files in {cache_dir_path} to {target_dir_path}")
        move_elements_across_dir(cache_dir_path, target_dir_path)
        try:
            os.rmdir(cache_dir_path)
        except FileNotFoundError:
            pass

        # Move File to Another dir
        print(f" > Finish dealing with file: {file_name}")
        used_pack_dir = os.path.join(pack_dir, "BOFTTPacks")
        if not os.path.isdir(used_pack_dir):
            os.mkdir(used_pack_dir)
        shutil.move(file_path, os.path.join(used_pack_dir, file_name))


def _rename_file_with_num(dir: str, file_name: str, input_num: int):
    file_path = os.path.join(dir, file_name)
    new_file_name = f"{input_num} {file_name}"
    new_file_path = os.path.join(dir, new_file_name)
    shutil.move(file_path, new_file_path)
    print(f"Rename {file_name} to {new_file_name}.")
    print()


def _set_file_num(
    dir: str,
    allow_ext: List[str] = [],
    disallow_ext: List[str] = [],
    allow_others: bool = True,
):
    file_names = []
    for file_name in os.listdir(dir):
        file_path = os.path.join(dir, file_name)
        # Not File?
        if not os.path.isfile(file_path):
            continue
        # Has been numbered?
        if file_name.split()[0].isdigit():
            continue
        # Linux: Has Partial File?
        part_file_path = f"{file_path}.part"
        if os.path.isfile(part_file_path):
            continue
        # Linux: Empty File?
        if os.path.getsize(file_path) == 0:
            continue
        # Is Allowed?
        file_ext = file_name.rsplit(".", 1)[-1]
        allowed = allow_others
        if file_ext in allow_ext:
            allowed = True
        elif file_ext in disallow_ext:
            allowed = False
        if not allowed:
            continue
        file_names.append(file_name)

    # Print Selections
    print(f"Here are files in {dir}:")
    for i, file_name in enumerate(file_names):
        print(f" - {i}: {file_name}")

    print("Input a number: to set num [0] to the first selection.")
    print("Input two numbers: to set num [1] to the selection in index [0].")
    input_str = input("Input:")
    input_str_split = input_str.split()
    if len(input_str_split) == 2:
        file_name = file_names[int(input_str_split[0])]
        input_num = int(input_str_split[1])
        _rename_file_with_num(dir, file_name, input_num)
    elif len(input_str_split) == 1:
        file_name = file_names[0]
        input_num = int(input_str_split[0])
        _rename_file_with_num(dir, file_name, input_num)
    else:
        print("Invaild input.")
        print()


def set_file_num(dir: str):
    while True:
        _set_file_num(
            dir,
            allow_ext=["zip", "7z", "rar", "mp4", "bms", "bme", "bml", "pms"],
            disallow_ext=[],
            allow_others=False,
        )


OPTIONS: List[Tuple[Callable, List[Tuple[InputType, str]]]] = [
    (
        set_file_num,
        [
            (InputType.Path, "Pack Dir"),
        ],
    ),
]

OPTIONS: List[Tuple[Callable, List[Tuple[InputType, str]]]] = [
    (
        unzip_numeric_to_bms_folder,
        [
            (InputType.Path, "Pack Dir"),
            (InputType.Path, "Cache Dir"),
            (InputType.Path, "Root Dir"),
        ],
    ),
    (
        unzip_with_name_to_bms_folder,
        [
            (InputType.Path, "Pack Dir"),
            (InputType.Path, "Cache Dir"),
            (InputType.Path, "Root Dir"),
        ],
    ),
    (
        set_file_num,
        [
            (InputType.Path, "RawFile Dir"),
        ],
    ),
]

File: cli/src/options/rawpack.rs
use log::info;
use smol::{fs, io, stream::StreamExt};
use std::path::Path;

use crate::fs::{
    is_dir_having_file,
    moving::{MoveOptions, move_elements_across_dir},
    rawpack::{
        get_num_set_file_names, move_out_files_in_folder_in_cache_dir, unzip_file_to_cache_dir,
    },
};

/// Extract numerically named pack files to BMS folders
pub async fn unzip_numeric_to_bms_folder(
    pack_dir: impl AsRef<Path>,
    cache_dir: impl AsRef<Path>,
    root_dir: impl AsRef<Path>,
    confirm: bool,
) -> io::Result<()> {
    let pack_dir = pack_dir.as_ref();
    let cache_dir = cache_dir.as_ref();
    let root_dir = root_dir.as_ref();

    if !cache_dir.exists() {
        fs::create_dir_all(cache_dir).await?;
    }
    if !root_dir.exists() {
        fs::create_dir_all(root_dir).await?;
    }

    let num_set_file_names = get_num_set_file_names(pack_dir)?;

    if confirm {
        for file_name in &num_set_file_names {
            info!(" --> {}", file_name);
        }
        info!("-> Confirm [y/N]:");
        // TODO: Implement user input confirmation
        return Ok(());
    }

    for file_name in num_set_file_names {
        let file_path = pack_dir.join(&file_name);
        let id_str = file_name.split(' ').next().unwrap_or("");

        // Prepare an empty cache dir
        let cache_dir_path = cache_dir.join(id_str);

        if cache_dir_path.exists() && is_dir_having_file(&cache_dir_path).await? {
            fs::remove_dir_all(&cache_dir_path).await?;
        }

        if !cache_dir_path.exists() {
            fs::create_dir_all(&cache_dir_path).await?;
        }

        // Unpack & Copy
        unzip_file_to_cache_dir(&file_path, &cache_dir_path).await?;

        // Move files in dir
        let move_result = move_out_files_in_folder_in_cache_dir(&cache_dir_path).await?;
        if !move_result {
            continue;
        }

        // Find Existing Target dir
        let mut target_dir_path = None;
        let mut entries = fs::read_dir(root_dir).await?;
        while let Some(entry) = entries.next().await {
            let entry = entry?;
            let dir_name = entry.file_name().to_string_lossy().into_owned();
            let dir_path = entry.path();

            if !entry.file_type().await?.is_dir() {
                continue;
            }

            if !(dir_name.starts_with(id_str)
                && (dir_name.len() == id_str.len() || dir_name[id_str.len()..].starts_with('.')))
            {
                continue;
            }
            target_dir_path = Some(dir_path);
        }

        // Create New Target dir
        let target_dir_path = if let Some(path) = target_dir_path {
            path
        } else {
            root_dir.join(id_str)
        };

        if !target_dir_path.exists() {
            fs::create_dir_all(&target_dir_path).await?;
        }

        // Move cache to bms dir
        info!(
            " > Moving files in {} to {}",
            cache_dir_path.display(),
            target_dir_path.display()
        );
        move_elements_across_dir(
            &cache_dir_path,
            &target_dir_path,
            MoveOptions::default(),
            Default::default(),
        )
        .await?;

        // Try to remove empty cache directory
        fs::remove_dir(&cache_dir_path).await.ok();

        // Move File to Another dir
        info!(" > Finish dealing with file: {}", file_name);
        let used_pack_dir = pack_dir.join("BOFTTPacks");
        if !used_pack_dir.exists() {
            fs::create_dir_all(&used_pack_dir).await?;
        }
        fs::rename(&file_path, used_pack_dir.join(&file_name)).await?;
    }

    Ok(())
}

/// Extract files with names to BMS folders
pub async fn unzip_with_name_to_bms_folder(
    pack_dir: impl AsRef<Path>,
    cache_dir: impl AsRef<Path>,
    root_dir: impl AsRef<Path>,
    confirm: bool,
) -> io::Result<()> {
    let pack_dir = pack_dir.as_ref();
    let cache_dir = cache_dir.as_ref();
    let root_dir = root_dir.as_ref();

    if !cache_dir.exists() {
        fs::create_dir_all(cache_dir).await?;
    }
    if !root_dir.exists() {
        fs::create_dir_all(root_dir).await?;
    }

    let mut num_set_file_names = Vec::new();
    let mut entries = fs::read_dir(pack_dir).await?;
    while let Some(entry) = entries.next().await {
        let entry = entry?;
        let file_name = entry.file_name().to_string_lossy().into_owned();
        if !entry.file_type().await?.is_file() {
            continue;
        }

        if file_name.ends_with(".zip") || file_name.ends_with(".7z") || file_name.ends_with(".rar")
        {
            num_set_file_names.push(file_name);
        }
    }

    if confirm {
        for file_name in &num_set_file_names {
            info!(" --> {}", file_name);
        }
        info!("-> Confirm [y/N]:");
        // TODO: Implement user input confirmation
        return Ok(());
    }

    for file_name in num_set_file_names {
        let file_path = pack_dir.join(&file_name);
        let file_name_without_ext = if let Some(dot_pos) = file_name.rfind('.') {
            &file_name[..dot_pos]
        } else {
            &file_name
        };

        let file_name_without_ext = file_name_without_ext.trim_end_matches('.');

        // Prepare an empty cache dir
        let cache_dir_path = cache_dir.join(file_name_without_ext);

        if cache_dir_path.exists() && is_dir_having_file(&cache_dir_path).await? {
            fs::remove_dir_all(&cache_dir_path).await?;
        }

        if !cache_dir_path.exists() {
            fs::create_dir_all(&cache_dir_path).await?;
        }

        // Unpack & Copy
        unzip_file_to_cache_dir(&file_path, &cache_dir_path).await?;

        // Move files in dir
        let move_result = move_out_files_in_folder_in_cache_dir(&cache_dir_path).await?;
        if !move_result {
            continue;
        }

        let target_dir_path = root_dir.join(file_name_without_ext);

        // Create New Target dir
        if !target_dir_path.exists() {
            fs::create_dir_all(&target_dir_path).await?;
        }

        // Move cache to bms dir
        info!(
            " > Moving files in {} to {}",
            cache_dir_path.display(),
            target_dir_path.display()
        );
        move_elements_across_dir(
            &cache_dir_path,
            &target_dir_path,
            MoveOptions::default(),
            Default::default(),
        )
        .await?;

        // Try to remove empty cache directory
        fs::remove_dir(&cache_dir_path).await.ok();

        // Move File to Another dir
        info!(" > Finish dealing with file: {}", file_name);
        let used_pack_dir = pack_dir.join("BOFTTPacks");
        if !used_pack_dir.exists() {
            fs::create_dir_all(&used_pack_dir).await?;
        }
        fs::rename(&file_path, used_pack_dir.join(&file_name)).await?;
    }

    Ok(())
}

/// Rename file with number
async fn _rename_file_with_num(
    dir: impl AsRef<Path>,
    file_name: &str,
    input_num: i32,
) -> io::Result<()> {
    let dir = dir.as_ref();
    let file_path = dir.join(file_name);
    let new_file_name = format!("{} {}", input_num, file_name);
    let new_file_path = dir.join(&new_file_name);

    fs::rename(&file_path, &new_file_path).await?;
    info!("Rename {} to {}.", file_name, new_file_name);
    info!("");

    Ok(())
}

/// Set file number (interactive)
pub async fn set_file_num(dir: impl AsRef<Path>) -> io::Result<()> {
    let dir = dir.as_ref();

    let mut file_names = Vec::new();
    let mut entries = fs::read_dir(dir).await?;

    while let Some(entry) = entries.next().await {
        let entry = entry?;
        let file_name = entry.file_name().to_string_lossy().into_owned();
        let file_path = entry.path();

        // Not File?
        if !entry.file_type().await?.is_file() {
            continue;
        }

        // Has been numbered?
        if file_name
            .split_whitespace()
            .next()
            .is_some_and(|s| s.chars().all(|c| c.is_ascii_digit()))
        {
            continue;
        }

        // Linux: Has Partial File?
        let part_file_path = format!("{}.part", file_path.display());
        if std::path::Path::new(&part_file_path).exists() {
            continue;
        }

        // Linux: Empty File?
        let metadata = fs::metadata(&file_path).await?;
        if metadata.len() == 0 {
            continue;
        }

        // Is Allowed?
        let file_ext = file_name.rsplit('.').next().unwrap_or("").to_lowercase();
        let allowed_exts = ["zip", "7z", "rar", "mp4", "bms", "bme", "bml", "pms"];
        let allowed = allowed_exts.contains(&file_ext.as_str());

        if !allowed {
            continue;
        }

        file_names.push(file_name);
    }

    // Print Selections
    info!("Here are files in {}:", dir.display());
    for (i, file_name) in file_names.iter().enumerate() {
        info!(" - {}: {}", i, file_name);
    }

    info!("Input a number: to set num [0] to the first selection.");
    info!("Input two numbers: to set num [1] to the selection in index [0].");
    info!("Input:");

    // TODO: Implement user input handling
    info!("Note: Interactive input not yet implemented");

    Ok(())
}
